File: src/main/java/frc/robot/subsystems/intake/IntakeIO.java
package frc.robot.subsystems.intake;

import org.littletonrobotics.junction.AutoLog;

public interface IntakeIO {
  // Cerealizer (n.) - 1. A device that dispenses cereal. 2. A typo the robot lead made that stuck.
  // fr tho this is hilarious
  @AutoLog
  public static class IntakeIOInputs {
    public double intakePositionDegrees = 0.0;
    public double intakeVelocityRpm = 0.0;
    public double intakeAppliedVolts = 0.0;
    public double[] intakeCurrentAmps = new double[] {};

    public double cerealizerPositionDegrees = 0.0;
    public double cerealizerVelocityRpm = 0.0;
    public double cerealizerAppliedVolts = 0.0;
    public double[] cerealizerCurrentAmps = new double[] {};

  }

  /** Updates the set of loggable inputs. */
  public default void updateInputs(IntakeIOInputs inputs) {}

  /** Run open loop at the specified voltage. */
  public default void set(double intakePercent, double cerealizerPercent) {}

  /** Run closed loop at the specified velocity. */
  public default void setVelocity(
      double intakeRpm, double cerealizerRpm) {}

  public default void setIntakeVelocity(double intakeRpm) {}
  public default void setCerealizerVelocity(double cerealizerRpm) {}

  public default void setCerealizer(double a) { }
  public default void setIntake(double a) {}
}

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt
package frc.robot.subsystems.intake

import edu.wpi.first.wpilibj.util.Color
import org.littletonrobotics.junction.AutoLog
import java.util.function.DoubleSupplier
import java.util.function.Supplier

/**
 *
 */
interface IntakeIO {
    /**
     *
     */// Cerealizer (n.) - 1. A device that dispenses cereal. 2. A typo the robot lead made that stuck.
    // fr tho this is hilarious
    @AutoLog
    open class IntakeIOInputs {
        /**
         *
         */
        @JvmField
        var intakePositionDegrees: Double = 0.0
        /**
         *
         */
        @JvmField
        var intakeVelocityRpm: Double = 0.0
        /**
         *
         */
        @JvmField
        var intakeAppliedVolts: Double = 0.0
        /**
         *
         */
        @JvmField
        var intakeCurrentAmps: DoubleArray = doubleArrayOf()

        /**
         *
         */
        @JvmField
        var cerealizerPositionDegrees: Double = 0.0
        /**
         *
         */
        @JvmField
        var cerealizerVelocityRpm: Double = 0.0
        /**
         *
         */
        @JvmField
        var cerealizerAppliedVolts: Double = 0.0

        /**
         *
         */
        @JvmField
        var cerealizerCurrentAmps: DoubleArray = doubleArrayOf()
    }

    /** Updates the set of loggable inputs.  */
    fun updateInputs(inputs: IntakeIOInputs) {}

    /** Run open loop at the specified voltage.  */
    operator fun set(intakePercent: Double, cerealizerPercent: Double) {}
    operator fun set(intakeMotorPercent: Double, intakePercent: Double, cerealizerPercent: Double) {}

    /** Run closed loop at the specified velocity.  */
    fun setVelocity(
        intakeRpm: Double, cerealizerRpm: Double
    ) {
    }

    /**
     *
     */
    fun setIntakeVelocity(intakeRpm: Double) {}
    /**
     *
     */
    fun setCerealizerVelocity(cerealizerRpm: Double) {}

    /**
     *
     */
    fun setCerealizer(a: Double) {}
    /**
     *
     */
    fun setIntake(a: Double) {}

    /**
     *
     */
    fun setIntakeVoltage(volts: Double) {}
    /**
     *
     */
    fun setCerealizerVoltage(volts: Double) {}

    /**
     *
     */
    val colorSensorProximity: DoubleSupplier

    /**
     *
     */
    val colorSensorRed: DoubleSupplier

    /**
     *
     */
    val colorSensorGreen: DoubleSupplier

    /**
     *
     */
    val colorSensorBlue: DoubleSupplier

    /**
     *
     */
    val colorSensorIR: DoubleSupplier

    /**
     *
     */
    val colorSensorColor: Supplier<Color>
}
